### PR TITLE
Helm ci test 

### PR DESCRIPTION
**Pull request by zhu733756**
**What type of PR is this?**
/kind test

**What this PR does / why we need it**:
Add another patch for #3418,  reusing the keadm e2e test codes to implement helm keadm e2e test. 

**Which issue(s) this PR fixes**:
Related to #3300 

**Special notes for your reviewer**:
See #3418

### CODE REVIEW

File: tests/e2e/scripts/keadm_e2e.sh [hunk 1]
```diff
@@ -35,7 +35,7 @@ function prepare_cluster() {
   kind create cluster --name test
 
   echo "wait the control-plane ready..."
-  kubectl wait --for=condition=Ready node/test-control-plane --timeout=60s
```
Comment (fisherxu) [inline tests/e2e/scripts/keadm_e2e.sh:38]: Why delete the timeout here? :)

File: tests/e2e/scripts/keadm_e2e.sh [hunk 2]
```diff
@@ -110,11 +110,16 @@ set -Ee
 trap cleanup EXIT
 trap cleanup ERR
 
+echo -e "\nBuild keadm..."
```
Comment (fisherxu) [inline tests/e2e/scripts/keadm_e2e.sh:113]:
```suggestion
echo -e "\nBuilding keadm..."
```

File: tests/e2e/scripts/keadm_e2e.sh [hunk 3]
```diff
@@ -110,11 +110,16 @@ set -Ee
...
 build_keadm
 
 export KUBECONFIG=$HOME/.kube/config
+
+echo -e "\nPrepare cluster..."
```
Comment (fisherxu) [inline tests/e2e/scripts/keadm_e2e.sh:118]:
```suggestion
echo -e "\nPreparing cluster..."
```

File: tests/e2e/scripts/keadm_e2e.sh [hunk 4]
```diff
@@ -110,11 +110,16 @@ set -Ee
...
 prepare_cluster
 
+echo -e "\nStart kubeedge..."
```
Comment (fisherxu) [inline tests/e2e/scripts/keadm_e2e.sh:121]:
```suggestion
echo -e "\nStarting KubeEdge..."
```

File: tests/e2e/scripts/keadm_e2e.sh [hunk 5]
```diff
@@ -110,11 +110,16 @@ set -Ee
...
 start_kubeedge
 
-run_test
+echo -e "\nRun test..."
```
Comment (fisherxu) [inline tests/e2e/scripts/keadm_e2e.sh:124]:
```suggestion
echo -e "\nRunning test..."
```

File: tests/e2e/scripts/keadm_e2e.sh [hunk 6]
```diff
@@ -110,11 +110,16 @@ set -Ee
...
+run_test
```
Comment (fisherxu) [inline tests/e2e/scripts/keadm_e2e.sh:125]:
```suggestion
run_test

```

File: tests/e2e/scripts/helm_keadm_e2e.sh [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+echo -e "\nBuild keadm..."
+build_keadm
+
+export KUBECONFIG=$HOME/.kube/config
+
+echo -e "\nPrepare cluster..."
+prepare_cluster
+
+echo -e "\nBuild cloud image..."
```
Comment (fisherxu) [inline tests/e2e/scripts/helm_keadm_e2e.sh:132]:
```suggestion
echo -e "\nBuilding cloud image..."
```

File: tests/e2e/scripts/helm_keadm_e2e.sh [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+build_cloud_image
+
+echo -e "\nStart kubeedge..."
+start_kubeedge
+
+echo -e "\nRun test..."
```
Comment (fisherxu) [inline tests/e2e/scripts/helm_keadm_e2e.sh:138]:
```suggestion
echo -e "\nRunning test..."
```

File: tests/e2e/scripts/helm_keadm_e2e.sh [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+export KUBECONFIG=$HOME/.kube/config
+
+echo -e "\nPrepare cluster..."
+prepare_cluster
+
+echo -e "\nBuild cloud image..."
+build_cloud_image
+
+echo -e "\nStart kubeedge..."
```
Comment (fisherxu) [inline tests/e2e/scripts/helm_keadm_e2e.sh:135]:
```suggestion
echo -e "\nStarting KubeEdge..."
```

File: tests/e2e/scripts/helm_keadm_e2e.sh [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+set -Ee
+trap cleanup EXIT
+trap cleanup ERR
+
+echo -e "\nBuild keadm..."
+build_keadm
+
+export KUBECONFIG=$HOME/.kube/config
+
+echo -e "\nPrepare cluster..."
```
Comment (fisherxu) [inline tests/e2e/scripts/helm_keadm_e2e.sh:129]:
```suggestion
echo -e "\nPreparing cluster..."
```

File: tests/e2e/scripts/helm_keadm_e2e.sh [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+      echo "Integration suite successfully passed all the tests !!"
+      exit 0
+  fi
+}
+
+set -Ee
+trap cleanup EXIT
+trap cleanup ERR
+
+echo -e "\nBuild keadm..."
```
Comment (fisherxu) [inline tests/e2e/scripts/helm_keadm_e2e.sh:124]:
```suggestion
echo -e "\nBuilding keadm..."
```

**Review by fisherxu — COMMENTED** [review]
Overall looks good, thanks for the great test :)

File: .github/workflows/main.yaml [hunk 12]
```diff
@@ -169,6 +169,45 @@ jobs:
...
+
+      - name: Install dependences
+        run: |
+          command -v ginkgo || go install github.com/onsi/ginkgo/ginkgo@latest
+          go install sigs.k8s.io/kind@v0.11.1
+          curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.19.3/bin/linux/amd64/kubectl && sudo install kubectl /usr/local/bin/kubectl
+          export RELEASE_VERSION=$(wget -qO - https://kubeedge.io/latestversion | cat)
+          sudo wget -qP /etc/kubeedge https://github.com/kubeedge/kubeedge/releases/download/${RELEASE_VERSION}/checksum_kubeedge-${RELEASE_VERSION}-linux-amd64.tar.gz.txt
+          sudo wget -qP /etc/kubeedge https://github.com/kubeedge/kubeedge/releases/download/${RELEASE_VERSION}/kubeedge-${RELEASE_VERSION}-linux-amd64.tar.gz
+          echo "$(cat /etc/kubeedge/checksum_kubeedge-${RELEASE_VERSION}-linux-amd64.tar.gz.txt) /etc/kubeedge/kubeedge-${RELEASE_VERSION}-linux-amd64.tar.gz" | sha512sum -c
```
Comment (gy95) [inline .github/workflows/main.yaml:197]: we should use the latest codes to do test, but not the latest release version to do test.

File: tests/e2e/scripts/helm_keadm_e2e.sh [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+KUBEEDGE_ROOT=$PWD
+WORKDIR=$(dirname $0)
+HELM_DIR=$KUBEEDGE_ROOT/build/helm
+E2E_DIR=$(realpath $(dirname $0)/..)
+IMAGE_TAG=$(date +%Y%m%d)
```
Comment (gy95) [inline tests/e2e/scripts/helm_keadm_e2e.sh:21]: this is not precise, if we trigger the CI twice in one day, Isn't this conflict with each other, IMHO, we'd better use some others like git commit info.

File: tests/e2e/scripts/helm_keadm_e2e.sh [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+  # ensure tokensecret is generated
+  while true; do
+      sleep 3
+      kubectl get secret -nkubeedge 2>/dev/null | grep -q tokensecret && break
+  done
+  
+  cd $KUBEEDGE_ROOT
+  export TOKEN=$(sudo _output/local/bin/keadm gettoken --kube-config=$KUBECONFIG)
+  sudo systemctl set-environment CHECK_EDGECORE_ENVIRONMENT="false"
+  sudo -E CHECK_EDGECORE_ENVIRONMENT="false" _output/local/bin/keadm join --token=$TOKEN --cloudcore-ipport=$MASTER_IP:10000 --edgenode-name=edge-node
```
Comment (gy95) [inline tests/e2e/scripts/helm_keadm_e2e.sh:73]: this join command will use the latest version release as edgecore, but not the latest codes. we'd better build a new release with the latest codes and use --kubeedge-version flag to specify which new release we will use.

File: .github/workflows/main.yaml (hunk 12 above)
Comment (gy95) [reply .github/workflows/main.yaml:197]: and I think this block is redundant, we donot need to download the release manually. Will not helm use images but not binary release?

**Comment by zhu733756** [comment]
Will be fixed.

**Comment by zhu733756** [comment]
@gy95 Thanks for your reply.
Since I can't find the reply places for each review question. I filled all as follows:
1. The cloud side deployed by helm would use images only, the edge side not changed.
2. I think the image tag doesn't matter, it is mainly used in the action test, and we also can use it by overwriting the variable,

The other questions will have a check to reply you.

**Comment by gy95** [comment]
> @gy95 Thanks for your reply. Since I can't find the reply places for each review question. I filled all as follows:
> 
> 1. The cloud side deployed by helm would use images only, the edge side not changed.
> 2. I think the image tag doesn't matter, it is mainly used in the action test, and we also can use it by overwriting the variable,
> 
> The other questions will have a check to reply you.

2. Yes, but what I worried about is that if we trigger two CI test concurrently, it will cause race condition :) maybe the two CI will use the same overwritten image. and it maybe cause confusion.
3. about using latest codes to do test. I've done some tries in https://github.com/kubeedge/kubeedge/pull/3469 and https://github.com/kubeedge/kubeedge/pull/3467. we can do this based on the two PRs. If OK, any ideas about the two PRs?

**Comment by fisherxu** [comment]
Keadm uses latest binary for edgecore is reasonable.  And in futrue, this test will replace the keadm-e2e test once keadm support to use helm. Can we merge this first, and do the `using latest binary for edgecore` in #3469 ?

**Comment by zhu733756** [comment]
> > @gy95 Thanks for your reply. Since I can't find the reply places for each review question. I filled all as follows:
> > 
> > 1. The cloud side deployed by helm would use images only, the edge side not changed.
> > 2. I think the image tag doesn't matter, it is mainly used in the action test, and we also can use it by overwriting the variable,
> > 
> > The other questions will have a check to reply you.
> 
> 2. Yes, but what I worried about is that if we trigger two CI test concurrently, it will cause race condition :) maybe the two CI will use the same overwritten image. and it maybe cause confusion.
> 3. about using latest codes to do test. I've done some tries in [[WIP]using lateset codes to do keadm_e2e #3469](https://github.com/kubeedge/kubeedge/pull/3469) and [add script for build release #3467](https://github.com/kubeedge/kubeedge/pull/3467). we can do this based on the two PRs. If OK, any ideas about the two PRs?

It is just used in kind, have not pushed outside, doesn't matter. The two CI tests I think it was in two isolated environments.

**Comment by zhu733756** [comment]
> Keadm uses latest binary for edgecore is reasonable. And in futrue, this test will replace the keadm-e2e test once keadm support to use helm. Can we merge this first, and do the `using latest binary for edgecore` in #3469 ?

Agree with you, let's move there to discuss it. @gy95 @fisherxu 
I think the mentioned points need to fix in @gy95 's prs.

File: .github/workflows/main.yaml (hunk 12 above)
Comment (fisherxu) [inline .github/workflows/main.yaml:197]: Ref: #3487

File: tests/e2e/scripts/helm_keadm_e2e.sh (hunk 14 above)
Comment (fisherxu) [reply tests/e2e/scripts/helm_keadm_e2e.sh:73]: We can just build a kubeedge release package and please it to /etc/kubeedge now..  but `/etc/kubeedge` is not a good place for the package.

**Review by fisherxu — COMMENTED** [review]
/approve

**Comment by fisherxu** [comment]
Need rebase..

**Comment by zhu733756** [comment]
> Need rebase..

Done.

**Comment by fisherxu** [comment]
Any comments? @gy95

File: tests/e2e/scripts/helm_keadm_e2e.sh [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+function prepare_cluster() {
+  kind create cluster --name test
+
+  echo "wait the control-plane ready..."
+  kubectl wait --for=condition=Ready node/test-control-plane --timeout=60s
+
+  kubectl create clusterrolebinding system:anonymous --clusterrole=cluster-admin --user=system:anonymous
+
+  # edge side don't support kind cni now, delete kind cni plugin for workaround
+  # kubectl delete daemonset kindnet -nkube-system
```
Comment (gy95) [inline tests/e2e/scripts/helm_keadm_e2e.sh:45]: this line is commented out? why.

File: tests/e2e/scripts/helm_keadm_e2e.sh (hunk 14 above)
Comment (zhu733756) [reply tests/e2e/scripts/helm_keadm_e2e.sh:73]: Currently, we only test the hostnetwork case, maybe need to test the external iptablesmgr, any ideas?

File: tests/e2e/scripts/helm_keadm_e2e.sh (hunk 15 above)
Comment (zhu733756) [reply tests/e2e/scripts/helm_keadm_e2e.sh:45]: In our history tests, it blocks the connection as the comment said "edge side don't support kind cni now". With cloudcore deployed with container type, I think the kind cni would not block it. In additional, I consider using the iptablemgr component outside, in that case, we would use service and the cloudcore will not share hostnetwork.

**Comment by gy95** [comment]
LGTM

File: build/helm/cloudcore/templates/rbac_cloudcore.yaml [hunk 16]
```diff
@@ -8,7 +8,7 @@ metadata:
   {{- end }}
 rules:
 - apiGroups: [""]
-  resources: ["nodes", "nodes/status", "configmaps", "pods", "pods/status", "secrets", "endpoints", "services","persistentvolumes","persistentvolumeclaims"]
+  resources: ["nodes",  "nodes/status", "serviceaccounts/token", "configmaps", "pods", "pods/status", "secrets", "endpoints", "services","persistentvolumes","persistentvolumeclaims"]
   verbs: ["get", "list", "watch", "create", "update"]
```
Comment (fisherxu) [inline build/helm/cloudcore/templates/rbac_cloudcore.yaml:12]: We still need `node` patch verb, can just help add ut here. Add a new resource as in https://github.com/kubeedge/kubeedge/pull/3466 or add verbs in L12 looks all fine, Add a new resource maybe better?

Comment (zhu733756) [reply build/helm/cloudcore/templates/rbac_cloudcore.yaml:12]: > 

OK, I will do the helm CRD patch. The CRDs in build/cloud dir will need @zivAnyvision‘s help

File: build/helm/cloudcore/templates/rbac_cloudcore.yaml [hunk 17]
```diff
@@ -34,6 +34,9 @@ rules:
 - apiGroups: ["apiextensions.k8s.io"]
   resources: ["customresourcedefinitions"]
   verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
+- apiGroups: ["networking.istio.io"]
```
Comment (zhu733756) [inline build/helm/cloudcore/templates/rbac_cloudcore.yaml:37]: Todo: maybe we need to do some combined tests with edgemesh.

**Comment by zhu733756** [comment]
Why e2e fail? Is it occurred by accident？ It seems to just add empty whitespace to the former comment with forced-push.

**Comment by fisherxu** [comment]
> Why e2e fail? Is it occurred by accident？ It seems to just add a whitespace to the former comment with forced-push.
> 
> See https://github.com/kubeedge/kubeedge/runs/4590380659?check_suite_focus=true

It should be occurred by accident I think.

**Comment by zhu733756** [comment]
> > Why e2e fail? Is it occurred by accident？ It seems to just add a whitespace to the former comment with forced-push.
> > See https://github.com/kubeedge/kubeedge/runs/4590380659?check_suite_focus=true
> 
> It should be occurred by accident I think.

I agree with you.

**Review by fisherxu — COMMENTED** [review]
/lgtm

**Pull request merged**